src/main/java/org/apache/pulsar/io/gcp/PubsubSource.java

```diff
@@ -52,16 +49,16 @@ public void open(Map<String, Object> config, SourceContext sourceContext) throws
             .build();
 
     Subscriber.Builder subscriberBuilder = Subscriber.newBuilder(projectSubscriptionName,
-        (PubsubMessage message, AckReplyConsumer consumer) -> {
-          Record<byte[]> record = new PubsubRecord(sourceContext.getOutputTopic(), message);
-          try {
-            queue.put(record);
-            consumer.ack();
-          } catch (InterruptedException e) {
-            consumer.nack();
-            throw new RuntimeException(e);
-          }
-        })
+            (PubsubMessage message, AckReplyConsumer consumer) -> {
+              Record<byte[]> record = new PubsubRecord(sourceContext.getOutputTopic(), message, consumer);
+              try {
+                consume(record);
+                record.ack();
```

Comment:
You shouldn't be calling `ack` here now. It will be called by Pulsar when the message is ready to be acknowledged.